### PR TITLE
allowed characters: string -> record

### PR DESCRIPTION
**Pull request by hediet**
Please only review https://github.com/microsoft/vscode/pull/138042/commits/093aa3c433320e9aaa32ca55fe8fc54e6fbe8336

This PR enables this scenario, which is already supported by the config system:

```json5
// user
"editor.unicodeHighlight.allowedCharacters": {
    "ο": true
}

// workspace
"[typescript]": {
    "editor.unicodeHighlight.allowedCharacters": {
        "γ": true
    }
},
"editor.unicodeHighlight.allowedCharacters": {
    "α": true
}
```

-> Resolved settings for `foo.ts`:
```json5
"editor.unicodeHighlight.allowedCharacters": {
    "ο": true,
    "γ": true
    "α": true,
}
```

### CODE REVIEW

File: src/vs/editor/common/config/commonEditorConfig.ts [hunk 1]
```diff
@@ -398,7 +383,7 @@ export abstract class CommonEditorConfiguration extends Disposable implements IC
 					continue;
 				}
 				if (baseValue && typeof baseValue === 'object' && subsetValue && typeof subsetValue === 'object') {
-					if (!this._subsetEquals(baseValue, subsetValue)) {
+					if (!objects.equals(baseValue, subsetValue)) {
```
Comment (alexdima) [inline src/vs/editor/common/config/commonEditorConfig.ts:386]: I don't think this change preserves existing semantics. Here's an example on how `_subsetEquals` was working before:

```
editor.updateOptions({ minimap: { enabled: true, showSlider: 'always' } });
editor.updateOptions({ minimap: { enabled: false } }); // effective value: { enabled: false, showSlider: 'always' }
editor.updateOptions({ minimap: { enabled: false } }); // effective value: { enabled: false, showSlider: 'always' }
```

`_subsetEquals` was supposed to make the 3rd call be a no-op. But if I understand correctly the proposed changes, the third call will now go through because `object.equals({ enabled: false, showSlider: 'always' }, { enabled: false })` returns `false`.

File: src/vs/editor/common/config/commonEditorConfig.ts [hunk 2]
```diff
@@ -415,10 +400,10 @@ export abstract class CommonEditorConfiguration extends Disposable implements IC
 			return;
 		}
 		const newOptions = deepCloneAndMigrateOptions(_newOptions);
-		if (CommonEditorConfiguration._subsetEquals(this._rawOptions, newOptions)) {
+		if (CommonEditorConfiguration._topLevelSubsetEquals(this._rawOptions, newOptions)) {
 			return;
 		}
-		this._rawOptions = objects.mixin(this._rawOptions, newOptions || {});
+		this._rawOptions = Object.assign(this._rawOptions, newOptions || {});
```
Comment (alexdima) [inline src/vs/editor/common/config/commonEditorConfig.ts:406]: This change also breaks existing semantics. Same example as above:
```
// by default minimap.showSlider is 'mouseover'
editor.updateOptions({ minimap: { enabled: true, showSlider: 'always' } });
editor.updateOptions({ minimap: { enabled: false } }); // effective value: { enabled: false, showSlider: 'always' }
```

If I understand correctly the proposed change, the effective value will now be `{ enabled: false }` instead of `{ enabled: false, showSlider: 'always' }`

**Review by alexdima — CHANGES_REQUESTED** [review]
I think the proposed changes break the way editor options work for users of `updateOptions` that only pass in partial objects with the option they want to change with each call (unlike our workbench, which always passes in complete objects).

**Comment by hediet** [comment]
Will investigate test failures

File: src/vs/editor/common/config/editorOptions.ts [hunk 3]
```diff
@@ -828,13 +835,45 @@ abstract class BaseEditorOption<K1 extends EditorOption, V> implements IEditorOp
...
+	if (typeof value !== 'object' || typeof update !== 'object' || !value || !update) {
+		return new ApplyUpdateResult(update, value !== update);
+	}
+	if (Array.isArray(value) || Array.isArray(update)) {
+		const arrayEquals = Array.isArray(value) && Array.isArray(update) && arrays.equals(value, update);
+		return new ApplyUpdateResult(update, arrayEquals);
+	}
+	let didChange = false;
+	for (let key in update) {
+		if ((update as T & object).hasOwnProperty(key)) {
```
Comment (alexdima) [inline src/vs/editor/common/config/editorOptions.ts:866]: Would this throw when configuring e.g. `"editor.minimap": null` in settings? We've had a few incidents like that and we had to protect that the editor is still usable despite that bad input.

**Review by alexdima — APPROVED** [review]
Nice! Looks good! I just left a comment about maybe being more defensive towards `null` values passed in `applyUpdate`. Needs some testing to make sure that configuring things as `null` doesn't break the editor.

File: src/vs/editor/common/config/editorOptions.ts (hunk 3 above)
Comment (hediet) [reply src/vs/editor/common/config/editorOptions.ts:866]: When value or update are falsey, the function does an early return (see line 858).
